pennylane/capture/base_interpreter.py

```diff
@@ -0,0 +1,415 @@
+"""
+
+import copy
+from functools import partial, wraps
+from typing import Callable
+
+import jax
+
+import pennylane as qml
+
+from .flatfn import FlatFn
+from .primitives import (
+    AbstractMeasurement,
+    AbstractOperator,
+    adjoint_transform_prim,
+    cond_prim,
+    ctrl_transform_prim,
+    for_loop_prim,
+    grad_prim,
+    jacobian_prim,
+    qnode_prim,
+    while_loop_prim,
+)
+
+
+def jaxpr_to_jaxpr(
+    interpreter: "PlxprInterpreter", jaxpr: "jax.core.Jaxpr", consts, *args
+) -> "jax.core.Jaxpr":
+    """A convenience uility for converting jaxpr to a new jaxpr via an interpreter."""
+
+    def f(*inner_args):
+        return interpreter.eval(jaxpr, consts, *inner_args)
+
+    return jax.make_jaxpr(f)(*args).jaxpr
+
+
+class PlxprInterpreter:
+    """A template base class for defining plxpr interpreters
+
+    **Examples:**
+
+    .. code-block:: python
+
+        import jax
+        from pennylane.capture import PlxprInterpreter
+
+        class SimplifyInterpreter(PlxprInterpreter):
+
+            def interpret_operation(self, op):
+                new_op = qml.simplify(op)
+                if new_op is op:
+                    # if new op isn't queued, need to requeue op.
+                    data, struct = jax.tree_util.tree_flatten(new_op)
+                    new_op = jax.tree_util.tree_unflatten(struct, data)
+                return new_op
+
+    Now the interpreter can be used to transform functions and jaxpr:
+
+    >>> interpreter = SimplifyInterpreter()
+    >>> def f(x):
+    ...     qml.RX(x, 0)**2
+    ...     qml.adjoint(qml.Z(0))
+    ...     return qml.expval(qml.X(0) + qml.X(0))
+    >>> simplified_f = interpreter(f)
+    >>> print(qml.draw(simplified_f)(0.5)
+    0: ──RX(1.00)──Z─┤  <2.00*X>
+    >>> jaxpr = jax.make_jaxpr(f)(0.5)
+    >>> interpreter.eval(jaxpr.jaxpr, [], 0.5)
+    [expval(2.0 * X(0))]
+
+    It will also preserve higher order primitives by default:
+
+    >>> def g(x):
+    ...     @qml.for_loop(3)
+    ...     def loop(i, x):
+    ...         qml.RX(x, 0) ** i
+    ...         return x
+    ...     loop(1.0)
+    ...     return qml.expval(qml.Z(0) + 3*qml.Z(0))
+    >>> jax.make_jaxpr(interpreter(g))(0.5)
+    { lambda ; a:f32[]. let
+        _:f32[] = for_loop[
+        jaxpr_body_fn={ lambda ; b:i32[] c:f32[]. let
+            d:f32[] = convert_element_type[new_dtype=float32 weak_type=True] b
+            e:f32[] = mul c d
+            _:AbstractOperator() = RX[n_wires=1] e 0
+            in (c,) }
+        n_consts=0
+        ] 0 3 1 1.0
+        f:AbstractOperator() = PauliZ[n_wires=1] 0
+        g:AbstractOperator() = SProd[_pauli_rep=4.0 * Z(0)] 4.0 f
+        h:AbstractMeasurement(n_wires=None) = expval_obs g
+    in (h,) }
+
+
+    """
+
+    _env: dict
+    _primitive_registrations: dict["jax.core.Primitive", Callable] = {}
+    _op_math_cache: dict
+
+    def __init_subclass__(cls) -> None:
+        cls._primitive_registrations = copy.copy(cls._primitive_registrations)
```

Comment:
Neat way of ensuring that registration does not leak between interpreters! :100: